File: openharmony/arkcompiler_runtime_core/static_core/assembler/context.cpp
/*
 * Copyright (c) 2021-2024 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <string_view>
#include "assembly-parser.h"
#include "utils/number-utils.h"

namespace ark::pandasm {

void Context::Make(const std::vector<ark::pandasm::Token> &t)
{
    err = {};

    insNumber = 0;

    tokens = t;

    number = 1;

    end = false;

    token = std::string_view(&*(tokens[number - 1].wholeLine.begin() + tokens[number - 1].boundLeft),
                             tokens[number - 1].boundRight - tokens[number - 1].boundLeft);

    id = this->tokens[number - 1].type;
}

size_t Context::Len() const
{
    return token.size();
}

bool Context::ValidateFoundedRegisterName(char c, size_t n) const
{
    std::string_view p = token;

    p.remove_prefix(1);

    if (p.empty() || (p.size() > 1 && p[0] == '0')) {
        return false;
    }

    if (c != 'a') {
        for (const auto &ch : p) {
            if (std::isdigit(ch) == 0) {
                return false;
            }
        }
    } else {
        if (ToNumber(p) > n) {
            return false;
        }
    }

    return true;
}

bool Context::ValidateRegisterName(char c, size_t n) const
{
    if (token[0] == c) {
        return ValidateFoundedRegisterName(c, n);
    }

    return false;
}

bool Context::ValidateParameterName(size_t numberOfParamsAlreadyIs) const
{
    if (numberOfParamsAlreadyIs >= MAX_DWORD) {
        return false;
    }

    if (token[0] == 'a') {
        std::string_view p = token;

        p.remove_prefix(1);

        if (ToNumber(p) == numberOfParamsAlreadyIs) {
            return true;
        }
    }

    return false;
}

std::string_view Context::GiveToken()
{
    return token;
}

Token::Type Context::Next()
{
    if (this->tokens.size() > number) {
        return this->tokens[number].type;
    }

    return this->tokens[number - 1].type;
}

void Context::UpSignOperation()
{
    signop = id;
}

Token::Type Context::WaitFor()
{
    return signop;
}

bool Context::Mask()
{
    return end;
}

bool Context::NextMask()
{
    if (end) {
        return true;
    }

    return this->tokens.size() < number + 1;
}

// NOLINTNEXTLINE(cert-dcl21-cpp)
Token::Type Context::operator++(int)
{
    Token::Type lastId = id;

    if (this->tokens.size() > number) {
        ++number;

        id = this->tokens[number - 1].type;

        token = std::string_view(&*(tokens[number - 1].wholeLine.begin() + tokens[number - 1].boundLeft),
                                 tokens[number - 1].boundRight - tokens[number - 1].boundLeft);
    } else {
        end = true;
    }

    return lastId;
}

Token::Type Context::operator++()
{
    if (this->tokens.size() > number) {
        ++number;

        id = this->tokens[number - 1].type;

        token = std::string_view(&*(tokens[number - 1].wholeLine.begin() + tokens[number - 1].boundLeft),
                                 tokens[number - 1].boundRight - tokens[number - 1].boundLeft);
    } else {
        end = true;
    }

    return id;
}

// NOLINTNEXTLINE(cert-dcl21-cpp)
Token::Type Context::operator--(int)
{
    Token::Type lastId = id;

    if (number > 1) {
        end = false;

        --number;

        id = this->tokens[number - 1].type;

        token = std::string_view(&*(tokens[number - 1].wholeLine.begin() + tokens[number - 1].boundLeft),
                                 tokens[number - 1].boundRight - tokens[number - 1].boundLeft);
    } else {
        end = false;
    }

    return lastId;
}

Token::Type Context::operator--()
{
    if (number > 1) {
        end = false;

        --number;

        id = this->tokens[number - 1].type;

        token = std::string_view(&*(tokens[number - 1].wholeLine.begin() + tokens[number - 1].boundLeft),
                                 tokens[number - 1].boundRight - tokens[number - 1].boundLeft);
    } else {
        end = false;
    }

    return id;
}

Token::Type Context::operator*()
{
    return id;
}

}  // namespace ark::pandasm

File: openharmony/arkui_ace_engine/frameworks/bridge/arkts_frontend/koala_mirror/ui2abc/libarkts/src/arkts-api/peers/Context.ts
/*
 * Copyright (c) 2024 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { ArktsObject } from "./ArktsObject"
import { Program } from "../../generated"
import { global } from "../static/global"
import { passString, passStringArray } from "../utilities/private"
import { KNativePointer, nullptr, KBoolean } from "@koalaui/interop"
import { Config } from "./Config"
import { filterSource, throwError } from "../../utils"
import { AstNode } from "./AstNode"

export class Context extends ArktsObject {
    constructor(peer: KNativePointer) {
        super(peer)
    }

    static createFromString(source: string): Context {
        if (!global.configIsInitialized()) {
            throw new Error(`Config not initialized`)
        }
        return new Context(
            global.es2panda._CreateContextFromString(
                global.config,
                passString(source),
                passString(global.filePath)
            )
        )
    }
    
    /** @deprecated Use {@link createCacheFromFile} instead */
    static createCacheContextFromFile(
        configPtr: KNativePointer,
        fileName: string,
        globalContextPtr: KNativePointer,
        isExternal: KBoolean
    ): Context {
        return new Context(
            global.es2panda._CreateCacheContextFromFile(configPtr, passString(fileName), globalContextPtr, isExternal)
        );
    }

    static createFromFile(filePath: string, configPath: string, stdlibPath: string, outputPath: string): Context | undefined {
        const config = Config.create([
            "",
            "--arktsconfig",
            configPath,
            '--extension',
            'ets',
            '--stdlib',
            stdlibPath,
            filePath,
            '--output',
            outputPath,
        ])
        return new Context(
            global.es2panda._CreateContextFromFile(
                config.peer,
                passString(filePath)
            )
        )
    }

    static createCacheFromFile(filePath: string, config: Config, globalContext: GlobalContext, isExternal: boolean) {
        return new Context(
            global.es2panda._CreateCacheContextFromFile(
                config.peer,
                passString(filePath),
                globalContext.peer,
                isExternal
            )
        )
    }

    static createContextGenerateAbcForExternalSourceFiles(
        filenames: string[]): Context {
        if (!global.configIsInitialized()) {
            throwError(`Config not initialized`);
        }
        return new Context(
            global.es2panda._CreateContextGenerateAbcForExternalSourceFiles(global.config, filenames.length, passStringArray(filenames))
        );
    }

    destroy() {
        if (this.peer != nullptr) {
            global.es2panda._DestroyContext(this.peer)
            this.peer = nullptr
        }
    }
    
    /** @deprecated */
    static destroyAndRecreate(ast: AstNode): Context {
        console.log('[TS WRAPPER] DESTROY AND RECREATE');
        const source = filterSource(ast.dumpSrc());
        global.es2panda._DestroyContext(global.context);
        global.compilerContext = Context.createFromString(source) as any; // Improve: commonize Context

        return new Context(global.context);
    }

    get program(): Program {
        return new Program(global.es2panda._ContextProgram(this.peer));
    }
}

export class GlobalContext extends ArktsObject {
    static create(
        config: Config, externalFileList: string[]
    ): GlobalContext {
        return new GlobalContext(
            global.es2panda._CreateGlobalContext(
                config.peer,
                passStringArray(externalFileList),
                externalFileList.length,
                false
            )
        )
    }

    constructor(peer: KNativePointer) {
        super(peer)
    }

    destroy() {
        if (this.peer != nullptr) {
            global.es2panda._DestroyGlobalContext(this.peer)
            this.peer = nullptr
        }
    }
}
